### DISCUSSION TITLE
Edge API route fails but returns a 404 response code

### DISCUSSION THREAD

**Discussion by anirudhsama** [vercel/next.js#62445] [Help]
### Summary

 have an API route that fails with error INTERNAL_EDGE_FUNCTION_INVOCATION_FAILED, but instead of returning a 504 http status code in the response, it is sending a 404 instead. It also seems like whenever that error occurs, the _error path is also called. The original API route is defined in the App directory with edge runtime. Any idea what is happening here?
![CleanShot 2024-02-23 at 11 55 07@2x](https://github.com/vercel/next.js/assets/832760/14671f12-4cce-4ba9-b7d3-7ad1f2a67b42)

### Additional information

```js
Operating System:
  Platform: darwin
  Arch: arm64
  Version: Darwin Kernel Version 23.3.0: Wed Dec 20 21:30:44 PST 2023; root:xnu-10002.81.5~7/RELEASE_ARM64_T6000
Binaries:
  Node: 20.10.0
  npm: 10.2.3
  Yarn: 4.0.2
  pnpm: N/A
Relevant Packages:
  next: 14.0.4
  eslint-config-next: 14.0.4
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.3.3
Next.js Config:
  output: N/A
```

**Comment by icyJoseph** [comment]
Does the function work locally though? I wonder if its missing some dependency, or even system dependency, and that's preventing it from being called at all. 

- https://vercel.com/docs/errors/INTERNAL_EDGE_FUNCTION_INVOCATION_FAILED

Did it ever work? What packages does it use?

**Reply by anirudhsama** [reply]
It works most of the time locally and in prod. But when it does fail, I’m not sure why a 404 is being returned rather than a 504. If you see in the screenshot, there is another API which fails and it is also returning a 404 instead of 500.

**Reply by icyJoseph** [reply]
Yeah, I am a bit confused too, I'd guess, that it crashes out before it can be discovered by the server? and then it is like, it doesn't exist? no idea honestly.

Have you contacted the Vercel support team? If not, please do, they can help you more than I can at this stage.

**Reply by anirudhsama** [reply]
They said it’s an issue with the code. Their response:

Looking at the logs, it seems this occurs when the Edge function doesn't send a response within 25 seconds. This in turns calls the /_error Serverless Function, which I believe is responsible for the 404 error code we see in the logs.

If you look at your screenshot, the logs for the Edge functions invocations are always followed by a log from a Serverless Function 25 seconds later, and the request path is identical.

Please refer to the Docs page linked below to implement custom error handling:

https://nextjs.org/docs/pages/building-your-application/routing/custom-error

We recommend consulting the Next.js community where a wealth of information and troubleshooting tips are available.

**Reply by anirudhsama** [reply]
I have a /pages/_error.jsx page, I wonder if that is causing an issue. The documentation doesn’t say anything about how API routes behave with the _error.js page.

**Reply by icyJoseph** [reply]
Aha, but is the issue then that your functions runs for too long? 

Yeah if there's an error in API routes, the _error page gets triggered. At least in Pages Router it does, just like reaching for a non existent API route, would yield the 404 not found page.

**Reply by anirudhsama** [reply]
Ok there is some issue then. 

1. The API definitely exists. 
2. It is defined in the app directory. 
3. It seems like when no response is sent for 25 seconds, it triggers the INVOCATION error. 
4. When that happens, it tries to call _error and the status code is sent as 404, either because _error in app directory doesn’t exist or it incorrectly classifies the INVOCATION error as 404 instead of 504.
